torch/csrc/Module.cpp

```diff
@@ -524,6 +529,58 @@ PyObject *THPModule_isEnabledXNNPACK(PyObject * /* unused */)
   else Py_RETURN_FALSE;
 }
 
+namespace {
+struct WatchdogThread {
+  WatchdogThread(const std::string& name_, int timeout): name(name_), thrd([this, timeout] {
+      using namespace std::chrono_literals;
+      std::unique_lock<std::mutex> lock(mutex);
+      cv.wait_for(lock, timeout *1s, [this] { return exitFlag; });
+      if (!exitFlag) {
+        std::cerr << "Test " << name << " takes longer than " << timeout << " seconds" <<std::endl;
+        std::exit(-1);
+      }
+      }) {}
+  ~WatchdogThread() {
+    {
+      std::lock_guard<std::mutex> guard(mutex);
+      exitFlag = true;
+    }
+    cv.notify_one();
+    if (thrd.joinable())
+      thrd.join();
+  }
+private:
+  std::string name;
+  std::mutex mutex;
+  std::condition_variable cv;
+  bool exitFlag = false;
+  // Thread must be the last member, otherwise it will not get properly initialized
```

Comment:
because you need the cv/mutex etc to be initialized before it, is that right? Subtle...

Reply:
Yes, both `mutex`, `cv` and `exitFlag` must be initialized before thread is created. (I knew about this, but actually it was a bug in my initial implementation and this race was hit half way thru `torch_jit` tests)